Docker.md

```diff
@@ -1,20 +1,37 @@
-## development docker-compose.yaml file
+# Docker guidelines
 
-### Publishing ports
+## `Dockerfile`
+
+When building an image that is running some services,
+use the [`EXPOSE` directive](https://docs.docker.com/engine/reference/builder/#expose),
+to document which ports the services are listening to.
+Some tools ([such as the nginx proxy companion](https://github.com/nginx-proxy/acme-companion#step-3---proxied-containers))
+use it to define a contract between different containers too.
+Not using this directive doesn’t prevent exposing those ports to the Docker network.
+
+## `docker-compose.yml`
 
-By default, containers within the same docker-compose network are able to connect to services running
-on other containers on any port by using the container name, as long as the container is listening
-on 0:0:0:0.
+### Exposing ports
 
-The [`EXPOSE` directive isn't needed](https://docs.docker.com/engine/reference/builder/#expose), although
+By default, containers within the same Docker Compose network are able to connect to services running
+on other containers on any port by using the container name, as long as the service running in the
+container is listening on `0.0.0.0` which usually is the default too.
+
+The [`expose` configuration option](https://docs.docker.com/compose/compose-file/compose-file-v3/#expose)
+just makes this exposure more visible, and is used by
 some tools ([such as the nginx proxy companion](https://github.com/nginx-proxy/acme-companion#step-3---proxied-containers))
-use it to define a contract between different containers. Use it if you think it will be necessary.
+use it to define a contract between different containers.
+Use it to document how your containers are interacting.
+
+### Publishing ports
 
-When publishing services in a docker-compose.yaml file, ports should be explicitly published to the
-loopback interface, 127.0.0.1 to prevent docker from automatically binding them to 0.0.0.0
-(all interfaces, open to the world).
-For example, when using postgres or elasticsearch, we might want to make them available to developers, 
-but in the case that we deploy on a public server for testing we don't want them to be exposed.
+When publishing the ports of a service in a `docker-compose.yml` file for development, those ports
+should be explicitly published to the loopback interface `127.0.0.1` of the host, to prevent docker
+from automatically binding them to `0.0.0.0` (all interfaces of the host, open to the world).
+Publish ports of a non-public service only if you think it will be necessary for development.
+For example, when using `postgres` or `elasticsearch`, we might want to make them available to
+developers, but in the case that we deploy on a public server for testing we don't want them to be
+open to the world.
```

Comment:
As you recently did with musicbrainz-docker, do we want to propose a standard to set an env variable in `.env` for the exposed bind host so that users can change this if necessary?

Reply:
In `musicbrainz-docker` it defaults to listening to all interfaces because it has to work in production out of the box. I’m not sure whether or not it would be useful for development-only Docker Compose projects, and even then it should probably default to `127.0.0.1:` instead. So I don’t see how it could be standardized and if that is very useful anyway.

Reply:
yes, you're right. For local development this isn't as important. We can skip it.